akka-stream/src/main/scala/akka/stream/scaladsl/Source.scala

```diff
@@ -204,6 +204,17 @@ object Source {
     })
 
   /**
+   * Create [[Source]] that will continually produce given elements in specified order.
+   *
+   * Start a new 'cycled' `Source` from the given elements. The producer stream of elements
+   * will continue infinitely by repeating the same sequence of elements provided during construction.
+   */
+  def cycle[T](f: () ⇒ Iterator[T]): Source[T, NotUsed] = fromIterator(() ⇒ {
+    if (f().isEmpty) Iterator.empty
```

Comment:
That is still evaluating `f` several times and in the end there is really no way around that since `Iterator.continually` will evaluate it each time, and `f` doesn't have to return same thing every time and any such evaluation that is `empty` is bad. I think we should fail for empty, i.e. 
```scala
val iter = Iterator.continually {val i = f(); if (i.isEmpty) throw new NoSuchElementException("empty iterator") else i}.flatten
```